modules/tests/src/main/kotlin/com/github/davenury/tests/Changes.kt

```diff
@@ -145,24 +167,67 @@ class OnePeersetChanges(
     private val sender: Sender,
 ) {
     private var parentId = AtomicReference(InitialHistoryEntry.getId())
-    private var consensusLeader: PeerId? = null
+    private var consensusLeader = AtomicReference<PeerAddress?>(null)
 
     suspend fun introduceChange(change: Change): ChangeState {
-        val address = peersAddresses.find { it.peerId == consensusLeader }!!
-        return sender.executeChange(
-            address,
-            change
-        )
+        return try {
+            sender.executeChange(
+                consensusLeader.get()!!,
+                change
+            )
+        } catch (e: Exception) {
+            logger.info("Consensus leader ${consensusLeader.get()} is dead, I'm trying to get a new one")
+            populateConsensusLeader()
+            return introduceChange(change)
+        }
+    }
+
+    suspend fun getChange(): Change {
+        return try {
+            httpClient.get("http://${consensusLeader.get()!!.address}/v2/last-change")
+        } catch (e: Exception) {
+            when (e) {
+                is ClientRequestException, is ServerResponseException -> throw e
+                else -> {
+                    logger.info("Consensus leader ${consensusLeader.get()} is dead, I'm trying to get a new one")
+                    populateConsensusLeader()
+                    return getChange()
+                }
+            }
+        }
     }
 
-    suspend fun getConsensusLeader(): PeerId =
-        sender.getConsensusLeaderId(peersAddresses.first())!!
-            .also { this.consensusLeader = it }
+    suspend fun populateConsensusLeader() {
+        consensusLeader.set(getConsensusLeader())
+    }
+
+    private suspend fun getConsensusLeader(peerAddresses: List<PeerAddress> = peersAddresses): PeerAddress {
+        val address = peersAddresses.firstOrNull() ?: throw IllegalStateException("I have no more peers to ask!")
+        val peerId = try {
+            sender.getConsensusLeaderId(address)
+        } catch (e: Exception) {
+            logger.info("$address is dead, I'm trying to get consensus leader from another one")
+            return getConsensusLeader(peerAddresses.filterNot { it == address })
```

Comment:
I think this will be more efficient
```suggestion
            return getConsensusLeader(peerAddresses.drop(1))
```